### DISCUSSION TITLE
Make imported Nodes from XML writable

### DISCUSSION THREAD

**Discussion by deretive** [FreeOpcUa/python-opcua#1379] [Q&A]
Hey, 
I try to make an information model that contains many nodes from predefined companion specifications. I access the already existing nodes through a reference to the specific node in my xml file. The server just builds its address space from this xml file, no nodes are added explicitly in the address space. 
Is it possible to make the imported Nodes writable, like it is for self defined nodes?

**Comment by AndreasHeine** [comment]
does the companion specifications allow the client write?

**Reply by deretive** [reply]
For the specific nodes? No, there is no access level set. they look like:
`  <UAVariable NodeId="ns=1;i=15583" BrowseName="DefaultRolePermissions" ParentNodeId="ns=1;i=15011" DataType="i=96" ValueRank="1" ArrayDimensions="0">`
`    <DisplayName>DefaultRolePermissions</DisplayName>
    <References>
      <Reference ReferenceType="HasTypeDefinition">i=68</Reference>
      <Reference ReferenceType="HasProperty" IsForward="false">ns=1;i=15011</Reference>
    </References>
  </UAVariable>`

**Reply by deretive** [reply]
So should I declare each node again in my address space and set it writable? or can I somehow do it in my xml file?

**Reply by oroulet** [reply]
you can either modify xml file to make them writable (probably not a good idea if this file is created by someone else and might be updated)
or get  link to node in you code and set it writable something ala:   my_node = self.server.get_node("ns=zzzz, i=x").set_writable(True)

**Reply by deretive** [reply]
Thanks!
So there is no option to make the HasComponent/HasProperty nodes writable? 
In this case the preferred way is to build my server is in the address space instead of in a nodeset2.xml that is imported?

**Reply by AndreasHeine** [reply]
HasComponent and HasProperty are References and defined by OPC UA Spec. !

**Reply by AndreasHeine** [reply]
the umati community has a sample server https://github.com/umati/Sample-Server-asyncio the models are build in UaModeler and then imported into the server

**Reply by deretive** [reply]
I think I formulated it a bit unlucky, sorry. I meant, if it is possible to create lets say an UAObject, that contains some variables. The variables are not explicitly declared in the xml, but implemented through references to already existing nodes. Means it uses the particular, already defined variable ManufacturerId from the Robotics nodeset2.xml with its attributes, like the description. 
But after having a look at the examples you gave me, I think the best option is to create a independent XML file, that only makes use of HasTypeDefinition references to reference a already existing Type and not to access single variables through references.
